### PR TITLE
[BALDE-15] Add Vector(x,y)

### PR DESCRIPTION
**Pull request by 4jeR**
Added basic Vec2 class

### CODE REVIEW

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: constexpr function have to be defined in header file

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: operator + can't modify object

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: operator + can't modify object

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp:28]: should return new object

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: why cant't we multiply by 0? and we should use 0.0f notation

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: operator * shouldn't modify object

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp:68]: should return a new object

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: constexpr function have to be defined in header file

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: we don't need a variable if we just want to return this, it should be just return sqrt...., also consider making it constexpr

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: we don't need that variable it should just return, consider constexpr

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: precompiled header should be included in source file, and only if you use libraries included there

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: could be constexpr

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: if we dont't store pointer and dynamic memory it is unnesessery

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: i would add a faw spaces here

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: we should overload operator "<<" it would be more handy

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: this has to be defined in this file

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: this has to be defined in this file

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: we should store floats

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: we should store floats

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

File: tests/test_vector2.cpp
Comment (BIGbadEL) [inline tests/test_vector2.cpp:9]: you should add more tests

File: src/math/vec2.h
Comment (Blinkuu) [inline src/math/vec2.h]: i think DotProduct() would be more suitable name

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp:28]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp:9]: done

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: it's unnecessary

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: it should be return only function

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: there should be universal function that  normalize vector

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: it should return angle in radians

File: src/math/vec2.h
Comment (BIGbadEL) [reply src/math/vec2.h]: ...

File: src/math/vec2.h
Comment (BIGbadEL) [inline src/math/vec2.h]: we have two functions that do exacly the same thing

File: tests/test_vector2.cpp
Comment (BIGbadEL) [inline tests/test_vector2.cpp]: xDDD

File: tests/test_vector2.cpp
Comment (BIGbadEL) [reply tests/test_vector2.cpp:9]: what about +=, *=

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp:68]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp:9]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: fixed

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp]: XD

Comment (4jeR) [reply tests/test_vector2.cpp]: DONE

File: src/math/vec2.h
Comment (piotrmoszkowicz) [inline src/math/vec2.h]: Fix formatting

File: src/math/vec2.h
Comment (piotrmoszkowicz) [inline src/math/vec2.h]: Fix formatting

File: src/math/vec2.h
Comment (piotrmoszkowicz) [inline src/math/vec2.h]: As far as I know it subtracts. I know, we can argue that subtraction is kind of addition, but let's change it :)

File: src/math/vec2.h
Comment (piotrmoszkowicz) [inline src/math/vec2.h]: Aren't*

File: src/math/vec2.h
Comment (piotrmoszkowicz) [inline src/math/vec2.h]: A bit better explanation of param - let's say other vector at least

File: src/math/vec2.cpp
Comment (piotrmoszkowicz) [inline src/math/vec2.cpp]: If it's not needed remove it before merge.

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: It should test that vector has correct X and Y, not it's length.

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: It should test that vector has correct X and Y, not it's length.

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: No need to add 2 after ComparingOperator. 
These names work like this:
TEST(NameOfAllOfTestScenarios, SingleScenarioName)

For example here it simply should be TEST(NotEqualsOperator, DifferentClassInstances)

I would personally also add tests, which expect to get false.

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Afaik, you can move A.Reverse() into EXPECT_EQ instead of using one more variable.

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: ASSERT_FLOAT_EQ

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: ASSERT_FLOAT_EQ

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Change to ASSERT_TRUE(A == B)

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Change to ASSERT_TRUE

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Why not move it into the EXPECT_EQ -> EXPECT_EQ(..., A-=B)

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Move *=

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Move +=

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Move A*3.0f

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Move A-B

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: Move A+B

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: ASSERT_FLOAT_EQ

File: tests/test_vector2.cpp
Comment (piotrmoszkowicz) [inline tests/test_vector2.cpp]: ASSERT_FLOAT_EQ

**Review by piotrmoszkowicz — CHANGES_REQUESTED** [review]
Before all change make reformat code on all new files (CTRL + ALT + L when file is opened). It should add all the spaces and fix some formatting bugs.

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp]: done

File: tests/test_vector2.cpp
Comment (4jeR) [reply tests/test_vector2.cpp]: done

File: src/math/vec2.h
Comment (piotrmoszkowicz) [reply src/math/vec2.h]: Very last thing here - last line doesn't need two * and should have a bit different indent

File: src/math/vec2.h
Comment (piotrmoszkowicz) [reply src/math/vec2.h]: Same as above

File: src/math/vec2.h
Comment (4jeR) [reply src/math/vec2.h]: done

**Review by piotrmoszkowicz — APPROVED** [review]
Very good job! Let's wait for @BIGbadEL decision and we can merge it :)

File: src/math/vec2.cpp
Comment (BIGbadEL) [inline src/math/vec2.cpp]: function pow is very slow it this case simple multiplication will be better

**Review by BIGbadEL — COMMENTED** [review]
it's good, now what left is optimization. Try to make everything that's possible constexpr

File: src/math/vec2.cpp
Comment (4jeR) [reply src/math/vec2.cpp]: done

**Review by BIGbadEL — APPROVED** [review]

**Pull request merged**
